File: algorithms/2_medium/S/00713_subarray_product_less_than_k/solution.cpp
class Solution {
public:
    static int numSubarrayProductLessThanK(const vector<int>& nums, const int k) {
        assert(k >= 0 && k <= 1'000'000);

        if (k <= 1) { return 0; }

        const size_t len = nums.size();
        size_t lo = 0;
        int prod = 1;
        size_t cnt = 0;

        for (size_t hi = 0; hi < len; hi += 1) {
            assert(nums[hi] >= 1 && nums[hi] <= 1000);
            prod *= nums[hi];

            while (prod >= k) {
                prod /= nums[lo];
                lo += 1;
            }

            cnt += (hi + 1) - lo;
        }

        return static_cast<int>(cnt);
    }
};

File: algorithms/2_medium/S/00713_subarray_product_less_than_k/solution.rs
impl Solution {
    pub fn num_subarray_product_less_than_k(nums: Vec<i32>, k: i32) -> i32 {
        assert!(k >= 0 && k <= 1_000_000);

        if k <= 1 { return 0; }

        let len = nums.len();
        let mut lo = 0_usize;
        let mut prod = 1_i32;
        let mut cnt = 0_usize;

        for hi in 0..len {
            assert!(nums[hi] >= 1 && nums[hi] <= 1000);
            prod *= nums[hi];

            while prod >= k {
                prod /= nums[lo];
                lo += 1;
            }

            cnt += (hi + 1) - lo;
        }

        cnt.try_into().unwrap()
    }
}
